onadata/libs/filters.py

```diff
@@ -22,14 +21,22 @@ def filter_queryset(self, request, queryset, view):
 class RowLevelObjectPermissionFilter(guardian_filters.ObjectPermissionsFilter):
     def filter_queryset(self, request, queryset, view):
         """
-        Return queryset as-is if user is anonymous or headers contain
-        a one-time authentication request token (validation of the token is
-        delegated to the permission class)
+        Return queryset as-is if user is anonymous or super user. Otherwise,
+        narrow down the queryset to what the user is allowed to see.
         """
-        if (
-            request.user.is_anonymous
-            or OneTimeAuthToken.is_signed_request(request)[0]
-        ):
+
+        # Queryset cannot be narrowed down for anonymous and super users because
+        # they do not have object level permissions (actually a superuser could
+        # have object level permissions but `ServiceAccountUser` does not).
+        # The queryset will return a larger subset of data even objects they
+        # are not allowed to see (or modify).
+        # It is required to allow:
+        # - anonymous user to see public data
+        # - ServiceAccountUser to take actions on all objects on behalf of the
+        #   real user who is making the call to the API.
+        # The permissions validation is handled by the permission classes and
+        # should deny access to forbidden data
```

Comment:
:pray:
(let's review said classes on a call)

Reply:
As per discussed, this filter backend class is only used in the DataViewset.
`XFormDataPermissions` is the class that handles permissions and denies access if needed.